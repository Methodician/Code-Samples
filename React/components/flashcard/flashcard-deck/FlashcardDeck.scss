@import 'styles/Variables.scss';
$footer-height: 8rem;

:local(.FlashcardDeck) {
  margin-top: $header-height;
  font-family: $default-roboto-font;

  .quiz-container {
    height: calc(100vh - #{$header-height} - #{$footer-height} - 1rem);
    margin-top: 1rem;
    text-align: center;
  }

  .footer-buttons-container {
    height: 8rem;
    width: 100%;
    position: fixed;
    bottom: 0;
    display: grid;
    grid-template-columns: 1.5fr 3fr 1.5fr;
    flex-direction: row;
    align-items: center;
    justify-content: space-around;
  }
}
