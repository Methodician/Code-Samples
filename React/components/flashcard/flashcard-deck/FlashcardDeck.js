/* eslint-disable no-console */
import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';

// REDUX CONNECTIONS
import { connect } from 'react-redux';
import {
  getFlashcardDeck,
  getFlashcardCurrentIndex,
  getFlashcardIsRevealed,
} from 'selectors/flashcardSelectors';
import {
  _loadFlashcardDeck,
  _incrementFlashcardCurrentIndex,
  _decrementFlashcardCurrentIndex,
  _revealFlashcard,
  _hideFlashcard,
  _clearFlashcardSession,
} from 'actions/flashcardActions';

// 3RD PARTY LIBS
import { MdChevronLeft as BackArrow } from 'react-icons/md';
import { IoMdClose as Close } from 'react-icons/io';

// PROJECT COMPONENTS
import PageHeader, {
  Themes as HeaderThemes,
} from 'components/page-header/PageHeader';
import PageHeaderButton, {
  Themes as ButtonThemes,
  Positions,
} from 'components/page-header/page-header-button/PageHeaderButton';
import PageHeaderTitle from 'components/page-header/page-header-title/PageHeaderTitle';
import ProgressBar from 'components/shared/progress-bar/ProgressBar';
import SignalButton from 'components/shared/UI/signal-button/SignalButton';
import BackdropPanel from 'components/shared/backdrop-panel/BackdropPanel';
import { QuizIcon } from 'components/practice/Practice';
import BackArrowButton from 'components/shared/UI/arrow-buttons/BackArrowButton';
import FlashcardBasic from 'components/practice/flashcard/flashcard-types/FlashcardBasic';

// MISC...
import { routes } from 'routes/routes';
import styles from './FlashcardDeck.scss';

export class FlashcardDeck extends PureComponent {
  static propTypes = {
    match: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    loadFlashcardDeck: PropTypes.func.isRequired,
    incrementFlashcardCurrentIndex: PropTypes.func.isRequired,
    decrementFlashcardCurrentIndex: PropTypes.func.isRequired,
    revealFlashcard: PropTypes.func,
    hideFlashcard: PropTypes.func,
    clearFlashcardSession: PropTypes.func,
    flashcardDeck: PropTypes.array,
    flashcardCurrentIndex: PropTypes.number,
    isFlashcardRevealed: PropTypes.bool,
  };

  async componentDidMount() {
    const { match, loadFlashcardDeck } = this.props;
    const { lessonId } = match.params || '';
    loadFlashcardDeck(lessonId);
  }

  componentWillUnmount() {
    const { clearFlashcardSession } = this.props;
    clearFlashcardSession();
  }

  handleSelectPracticeAudioSpeed = setting => {
    this.setState({ isPracticeAudioSpeedRegular: setting });
  };

  handleNextClick = () => {
    const {
      incrementFlashcardCurrentIndex,
      flashcardCurrentIndex,
      flashcardDeck,
      hideFlashcard,
    } = this.props;
    hideFlashcard();
    incrementFlashcardCurrentIndex(flashcardCurrentIndex, flashcardDeck.length);
  };

  handlePreviousCardClick = () => {
    const {
      decrementFlashcardCurrentIndex,
      flashcardCurrentIndex,
      hideFlashcard,
    } = this.props;

    hideFlashcard();
    decrementFlashcardCurrentIndex(flashcardCurrentIndex);
  };

  handleRevealClick = () => {
    const { revealFlashcard } = this.props;
    revealFlashcard();
  };

  handleExitClicked = () => {
    const { lessonId } = this.props.match.params;
    return this.props.history.push(routes.PRACTICE.linkTo(lessonId));
  };

  handleQuizClicked = () => {
    const { lessonId } = this.props.match.params;
    return this.props.history.push(routes.QUIZ.linkTo(lessonId));
  };

  render() {
    const {
      flashcardDeck,
      flashcardCurrentIndex,
      isFlashcardRevealed,
    } = this.props;

    if (!flashcardDeck) return null;
    const deckItem = flashcardDeck[flashcardCurrentIndex];
    if (!deckItem) return null; // ideally we shouldn't be able to have an invalid deck index... can we?

    const isLastCard = flashcardCurrentIndex === flashcardDeck.length - 1;

    const headerText = isLastCard
      ? 'Final Card'
      : `Flashcards ${flashcardCurrentIndex + 1}/${flashcardDeck.length}`;

    let footerButton = null;
    if (isFlashcardRevealed) {
      if (isLastCard) {
        footerButton = (
          <SignalButton click={this.handleQuizClicked} buttonType="neutral">
            <>Start Quiz</>
          </SignalButton>
        );
      } else {
        footerButton = (
          <SignalButton click={this.handleNextClick} buttonType="neutral">
            <>Next</>
          </SignalButton>
        );
      }
    } else {
      footerButton = (
        <SignalButton click={this.handleRevealClick} buttonType="neutral">
          <>Reveal</>
        </SignalButton>
      );
    }

    return (
      <section className={styles.FlashcardDeck}>
        <PageHeader theme={HeaderThemes.BEGINNER}>
          <Fragment>
            <PageHeaderButton
              theme={ButtonThemes.LIGHT}
              Icon={Close}
              click={this.handleExitClicked}
              position={Positions.LEFT}
            />
            <PageHeaderTitle headerText={headerText} />
            <PageHeaderButton
              click={this.handleQuizClicked}
              theme={ButtonThemes.LIGHT}
              Icon={QuizIcon}
              position={Positions.RIGHT}
            />
          </Fragment>
        </PageHeader>
        <ProgressBar
          status={flashcardCurrentIndex + 1}
          length={flashcardDeck.length}
        />
        <main className="quiz-container">
          <BackdropPanel hasPadding={false}>
            {deckItem && <FlashcardBasic deckItem={deckItem} />}
          </BackdropPanel>
        </main>
        <footer className="footer-buttons-container">
          <BackArrowButton
            Icon={BackArrow}
            click={this.handlePreviousCardClick}
          />
          {footerButton}
        </footer>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  flashcardDeck: getFlashcardDeck(state),
  flashcardCurrentIndex: getFlashcardCurrentIndex(state),
  isFlashcardRevealed: getFlashcardIsRevealed(state),
});

const mapDispatchToProps = dispatch => ({
  loadFlashcardDeck: lessonId => dispatch(_loadFlashcardDeck(lessonId)),
  incrementFlashcardCurrentIndex: (flashcardCurrentIndex, deckLength) =>
    dispatch(
      _incrementFlashcardCurrentIndex(flashcardCurrentIndex, deckLength)
    ),
  decrementFlashcardCurrentIndex: flashcardCurrentIndex =>
    dispatch(_decrementFlashcardCurrentIndex(flashcardCurrentIndex)),
  revealFlashcard: () => dispatch(_revealFlashcard()),
  hideFlashcard: () => dispatch(_hideFlashcard()),
  clearFlashcardSession: () => dispatch(_clearFlashcardSession()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FlashcardDeck);
