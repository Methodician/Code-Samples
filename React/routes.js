// This is a good and proper way to manage routing in React
import Explore from 'components/explore/Explore';
import Learn from 'components/learn/Learn';
import Me from 'components/me/Me';
import Review from 'components/review/Review';
import Practice from 'components/practice/Practice';
import Quiz from 'components/quiz/Quiz';
import QuizResults from 'components/quiz/quiz-results/QuizResults';
import Lesson from 'components/lessons/lesson/Lesson';
import FlashcardDeck from 'components/practice/flashcard/flashcard-deck/FlashcardDeck';
import Unit from 'components/unit/Unit.js';
import UnitNav from 'components/learn/unit-nav/UnitNav';

export const routes = {
  EXPLORE: {
    path: '/explore',
    component: Explore,
    linkTo: () => '/explore',
  },
  LEARN: {
    path: '/learn',
    component: Learn,
    linkTo: () => '/learn',
  },
  COURSES: {
    path: '/courses/:courseId',
    component: UnitNav,
    linkTo: courseId => `/courses/${courseId}`,
  },
  ME: {
    path: '/me',
    component: Me,
    linkTo: () => '/me',
  },
  REVIEW: {
    path: '/review',
    component: Review,
    linkTo: () => '/review',
  },
  LESSON: {
    path: '/lessons/:lessonId',
    component: Lesson,
    linkTo: lessonId => `/lessons/${lessonId}`,
  },
  PRACTICE: {
    path: '/lessons/:lessonId/practice',
    component: Practice,
    linkTo: lessonId => `/lessons/${lessonId}/practice`,
  },
  QUIZ: {
    path: '/lessons/:lessonId/quiz',
    component: Quiz,
    linkTo: lessonId => `/lessons/${lessonId}/quiz`,
  },
  QUIZ_RESULTS: {
    path: '/lessons/:lessonId/quiz-results',
    component: QuizResults,
    linkTo: lessonId => `/lessons/${lessonId}/quiz-results`,
  },
  FLASHCARD_DECK: {
    path: '/lessons/:lessonId/flashcard-deck',
    component: FlashcardDeck,
    linkTo: lessonId => `/lessons/${lessonId}/flashcard-deck`,
  },
  UNIT: {
    path: '/units/:unitId',
    component: Unit,
    linkTo: unitId => `/units/${unitId}`,
  },
};
