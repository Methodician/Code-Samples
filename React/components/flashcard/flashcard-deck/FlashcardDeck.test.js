import React from 'react';
import { shallow } from 'enzyme';
import { FlashcardDeck } from './FlashcardDeck';
import flashcards from 'services/fixtures/unit2Lesson1Flashcards.json';

const _loadFlashcardDeck = jest.fn();
const _incrementFlashcardCurrentIndex = jest.fn();
const _decrementFlashcardCurrentIndex = jest.fn();
const match = {};
const history = {};

const component = (
  <FlashcardDeck
    loadFlashcardDeck={_loadFlashcardDeck}
    incrementFlashcardCurrentIndex={_incrementFlashcardCurrentIndex}
    decrementFlashcardCurrentIndex={_decrementFlashcardCurrentIndex}
    match={match}
    history={history}
    flashcardDeck={flashcards}
    flashcardCurrentIndex={2}
    revealFlashcard={jest.fn()}
    hideFlashcard={jest.fn()}
  />
);
const wrapper = shallow(component);

describe('FlashcardDeck component', () => {
  it('should render and match snapshots', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
