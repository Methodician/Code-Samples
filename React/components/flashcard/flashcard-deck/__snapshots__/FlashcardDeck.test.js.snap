// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashcardDeck component should render and match snapshots 1`] = `
<section
  className="FlashcardDeck"
>
  <PageHeader
    theme="beginner-theme"
  >
    <PageHeaderButton
      Icon={[Function]}
      click={[Function]}
      position="left"
      theme="light"
    />
    <PageHeaderTitle
      headerText="Flashcards 3/11"
    />
    <PageHeaderButton
      Icon={[Function]}
      click={[Function]}
      position="right"
      theme="light"
    />
  </PageHeader>
  <ProgressBar
    length={11}
    status={3}
  />
  <main
    className="quiz-container"
  >
    <BackdropPanel
      hasPadding={false}
    >
      <Connect(FlashcardBasic)
        deckItem={
          Object {
            "CardID": "BCC-002-01-003",
            "English1": "three",
            "English2": "",
            "Image": "",
            "Multiple": "",
            "Normal": "BCC-002-01-003-N",
            "Notes": "",
            "Pinyin": "sān",
            "PracticeType": "B",
            "RelatedCardID": "",
            "Simplified": "三",
            "Slow": "BCC-002-01-004-S",
            "Traditional": "三",
            "Word": "W",
          }
        }
      />
    </BackdropPanel>
  </main>
  <footer
    className="footer-buttons-container"
  >
    <BackArrowButton
      Icon={[Function]}
      click={[Function]}
    />
    <SignalButton
      buttonType="neutral"
      click={[Function]}
    >
      Reveal
    </SignalButton>
  </footer>
</section>
`;
